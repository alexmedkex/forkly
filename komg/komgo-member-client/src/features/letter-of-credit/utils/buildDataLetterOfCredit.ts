import { IDataLetterOfCredit, ICargo, ITrade, IDataLetterOfCreditBase } from '@komgo/types'
import { Counterparty } from '../../counterparties/store/types'
import { IMember } from '../../members/store/types'
import { findCounterpartyByStatic, findMembersByStatic } from '../../letter-of-credit-legacy/utils/selectors'
import { ImmutableObject } from '../../../utils/types'

interface BuildDataLetterOfCreditProps {
  cargo?: ICargo
  trade: ImmutableObject<ITrade>
  applicant: IMember
  beneficiary: IMember
  issuingBanks: Counterparty[]
  beneficiaryBanks?: IMember[]
  dataLetterOfCreditBase: IDataLetterOfCreditBase
}
export const buildDataLetterOfCredit = ({
  cargo,
  trade,
  applicant,
  beneficiary,
  issuingBanks,
  beneficiaryBanks,
  dataLetterOfCreditBase
}: BuildDataLetterOfCreditProps): Partial<IDataLetterOfCredit> => ({
  cargo,
  trade: trade.toJS(),
  applicant: applicant as any,
  beneficiary: beneficiary as any,
  issuingBank: findCounterpartyByStatic(issuingBanks, dataLetterOfCreditBase.issuingBank.staticId) as any,
  beneficiaryBank: dataLetterOfCreditBase.beneficiaryBank
    ? (findMembersByStatic(beneficiaryBanks, dataLetterOfCreditBase.beneficiaryBank.staticId) as any)
    : undefined,
  amount: dataLetterOfCreditBase.amount,
  issuingBankReference: dataLetterOfCreditBase.issuingBankReference,
  currency: dataLetterOfCreditBase.currency,
  expiryDate: dataLetterOfCreditBase.expiryDate
})
