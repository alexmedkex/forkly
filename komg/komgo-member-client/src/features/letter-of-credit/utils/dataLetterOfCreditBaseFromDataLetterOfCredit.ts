import { IDataLetterOfCredit, IDataLetterOfCreditBase } from '@komgo/types'

export const buildDataLetterOfCreditBase = (input: IDataLetterOfCredit): IDataLetterOfCreditBase => {
  return {
    version: input.version,
    applicant: {
      staticId: input.applicant.staticId
    },
    issuingBank: {
      staticId: input.issuingBank.staticId
    },
    beneficiary: {
      staticId: input.beneficiary.staticId
    },
    beneficiaryBank: input.beneficiaryBank ? { staticId: input.beneficiaryBank.staticId } : undefined,
    trade: {
      source: input.trade.source,
      sourceId: input.trade.sourceId
    },
    cargo: input.cargo
      ? {
          source: input.cargo.source,
          cargoId: input.cargo.cargoId,
          sourceId: input.cargo.sourceId
        }
      : undefined,
    amount: input.amount,
    expiryDate: input.expiryDate,
    currency: input.currency,
    issuingBankReference: input.issuingBankReference || ''
  }
}
