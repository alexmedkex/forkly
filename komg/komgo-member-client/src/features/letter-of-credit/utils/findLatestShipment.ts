import { IDataLetterOfCredit, ILetterOfCredit } from '@komgo/types'

export const findLatestShipment = (letter: ILetterOfCredit<IDataLetterOfCredit>) => {
  const { trade } = letter.templateInstance.data
  return (trade && trade.deliveryPeriod && trade.deliveryPeriod.endDate) || ''
}
