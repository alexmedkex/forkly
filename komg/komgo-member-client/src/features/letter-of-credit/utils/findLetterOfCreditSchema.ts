import {
  TRADE_SCHEMA as TRADE_SCHEMA_V2,
  CARGO_SCHEMA as CARGO_SCHEMA_V2,
  COMPANY_BASE_SCHEMA,
  DATA_LETTER_OF_CREDIT_BASE_SCHEMA,
  DATA_LETTER_OF_CREDIT_TEMPLATE_SCHEMA,
  findSchemaById
} from '@komgo/types'

const KNOWN_SCHEMAS = [
  TRADE_SCHEMA_V2,
  CARGO_SCHEMA_V2,
  COMPANY_BASE_SCHEMA,
  DATA_LETTER_OF_CREDIT_BASE_SCHEMA,
  DATA_LETTER_OF_CREDIT_TEMPLATE_SCHEMA
]

export const findLetterOfCreditSchema = (schemaId: string) => {
  const schema = findSchemaById(KNOWN_SCHEMAS, schemaId)
  if (!schema) {
    throw Error(`No schema found for ${schemaId}`)
  }
  return schema
}
