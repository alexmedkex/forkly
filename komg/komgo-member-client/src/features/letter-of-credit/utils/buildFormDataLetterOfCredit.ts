import { ICargo, ITrade, IDataLetterOfCreditBase } from '@komgo/types'
import { IMember } from '../../members/store/types'
import { ImmutableObject } from '../../../utils/types'

interface BuildDataLetterOfCreditBaseForCreateProps {
  cargo?: ICargo
  trade: ImmutableObject<ITrade>
  applicant: IMember
  beneficiary: IMember
}
export const buildFormDataLetterOfCredit = ({
  cargo,
  trade,
  applicant,
  beneficiary
}: BuildDataLetterOfCreditBaseForCreateProps): IDataLetterOfCreditBase => {
  const startAmount = parseFloat(Number(trade.get('quantity') * trade.get('price')).toFixed())

  return {
    version: 1,
    amount: !isNaN(startAmount) ? startAmount : 0.0,
    currency: trade.get('currency'),
    trade: { sourceId: trade.get('sourceId'), source: trade.get('source') },
    applicant: { staticId: applicant.staticId },
    beneficiary: { staticId: beneficiary.staticId },
    issuingBank: { staticId: '' },
    expiryDate: 'Choose a valid expiry date',
    cargo: cargo && { source: cargo.source, sourceId: cargo.sourceId, cargoId: cargo.cargoId }
  }
}
