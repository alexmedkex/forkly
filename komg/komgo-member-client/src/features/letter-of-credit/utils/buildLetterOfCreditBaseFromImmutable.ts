import { ILetterOfCredit, IDataLetterOfCredit, ILetterOfCreditBase, IDataLetterOfCreditBase } from '@komgo/types'
import { ImmutableObject } from '../../../utils/types'

export const buildLetterOfCreditBaseFromImmutable = (
  letterOfCredit: ImmutableObject<ILetterOfCredit<IDataLetterOfCredit>>,
  data: IDataLetterOfCreditBase,
  templateModel: any
): ILetterOfCreditBase<IDataLetterOfCreditBase> => {
  return {
    type: letterOfCredit.get('type'),
    version: letterOfCredit.get('version'),
    templateInstance: {
      version: letterOfCredit.get('templateInstance').get('version'),
      templateStaticId: letterOfCredit.get('templateInstance').get('templateStaticId'),
      template: templateModel,
      templateSchemaId: letterOfCredit.get('templateInstance').get('templateSchemaId'),
      data,
      dataSchemaId: letterOfCredit.get('templateInstance').get('dataSchemaId'),
      bindings: letterOfCredit
        .get('templateInstance')
        .get('bindings')
        .toJS()
    }
  }
}
