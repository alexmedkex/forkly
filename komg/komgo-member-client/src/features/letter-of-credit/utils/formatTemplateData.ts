import { ITrade, IDataLetterOfCredit, ICompany } from '@komgo/types'
import { displayDate, dateFormats } from '../../../utils/date'
import Numeral from 'numeral'
import { capitalize, sentenceCaseWithAcronyms } from '../../../utils/casings'
import { type } from 'os'

const displayAmount = (input: number) => Numeral(input).format('0,0.00')

export const formatTemplateData = (data: Partial<IDataLetterOfCredit>): Partial<IDataLetterOfCredit> => ({
  ...data,
  trade: formatTrade(data.trade),
  amount: displayAmount(data.amount) as any,
  expiryDate: displayDate(data.expiryDate, dateFormats.inputs),
  beneficiaryBank: data.beneficiaryBank ? data.beneficiaryBank : ({ x500Name: { CN: 'No advising bank' } } as ICompany)
})

export const walker = (obj: any, replacer, parent?) => {
  if (!obj) {
    return null
  }
  return Array.isArray(obj)
    ? obj.map(item => {
        return typeof item === 'object' ? walker(item, replacer, parent) : item && replacer(parent, item)
      })
    : Object.entries(obj).reduce((memo, [key, value]) => {
        return {
          ...memo,
          [key]: typeof value === 'object' ? walker(value, replacer, key) : value && replacer(key, value)
        }
      }, {})
}

export const formatTrade = (trade: ITrade): ITrade => ({
  ...walker(trade, (key, value) => {
    switch (key) {
      case 'dealDate':
      case 'contractDate':
      case 'startDate':
      case 'endDate':
        return displayDate(value, dateFormats.inputs)
      case 'law':
        const [place] = value.split('LAW')
        return `${capitalize(place)}law` // e.g. New York Law
      case 'price':
        return displayAmount(trade.price)
      default:
        // check for enums and uppercase e.g. DAYS or TO_BE_FINANCED
        return typeof value === 'string' && value.match(/^([A-Z]|_)*$/) ? sentenceCaseWithAcronyms(value) : value
    }
  })
})
