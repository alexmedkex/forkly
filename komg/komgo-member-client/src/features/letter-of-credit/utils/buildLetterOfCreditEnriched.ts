import { IDataLetterOfCredit, ILetterOfCredit } from '@komgo/types'
import { findRole } from './findRole'
import { findTasksByLetterOfCreditStaticId } from './taskUtil'
import { TaskStatus, Task } from '../../tasks/store/types'
import { findLatestShipment } from './findLatestShipment'
import { ILetterOfCreditEnriched } from '../store/types'

export const buildLetterOfCreditEnriched = (
  letter: ILetterOfCredit<IDataLetterOfCredit>,
  tasks: Task[],
  companyStaticId: string
) => {
  const role = findRole(letter, companyStaticId)
  const letterTasks = findTasksByLetterOfCreditStaticId(tasks, letter.staticId)
  const enriched: ILetterOfCreditEnriched = {
    ...letter,
    role,
    actionStatus: letterTasks.length ? TaskStatus.ToDo : TaskStatus.Done,
    latestShipment: findLatestShipment(letter),
    tasks: letterTasks
  }
  return enriched
}
