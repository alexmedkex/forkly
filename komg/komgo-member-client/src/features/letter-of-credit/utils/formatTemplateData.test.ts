import { buildFakeTrade, buildFakeCargo, buildFakeDataLetterOfCredit, Law } from '@komgo/types'
import { formatTemplateData, walker, formatTrade } from './formatTemplateData'
import { sentenceCaseWithAcronyms } from '../../../utils/casings'

describe('formatTemplateData', () => {
  it('matches snapshot', () => {
    const trade = {
      ...buildFakeTrade({ version: 2 }),
      law: Law.NewYorkLaw,
      dealDate: '2017-12-31T00:00:00.000Z',
      contractDate: '2017-12-01T00:00:00.000Z'
    }
    const cargo = buildFakeCargo()
    const amount = 100
    const expiryDate = '2019-08-21T15:47:51.817Z'

    const data = { ...buildFakeDataLetterOfCredit(), expiryDate, trade, cargo, amount }

    expect(formatTemplateData(data)).toMatchSnapshot()
  })
})

describe('walker', () => {
  const replacer = (key, value) => sentenceCaseWithAcronyms(value)

  it('patch all the props', () => {
    const obj = { foo: 'FOO', bar: 'BAR' }
    expect(walker(obj, replacer)).toEqual({ foo: 'Foo', bar: 'Bar' })
  })

  it('patch all the props of an object with leaves', () => {
    const obj = { foo: 'FOO', deep: { bar: 'BAR' } }
    expect(walker(obj, replacer)).toEqual({ foo: 'Foo', deep: { bar: 'Bar' } })
  })

  it('patch all the props of an object with leaves that has an array of object', () => {
    const obj = { foo: 'FOO', deep: [{ bar: 'BAR' }] }
    expect(walker(obj, replacer)).toEqual({ foo: 'Foo', deep: [{ bar: 'Bar' }] })
  })

  it('patch all the props of an object with leaves that has an array of simple types', () => {
    const obj = { foo: 'FOO', deep: ['BAR'] }
    expect(walker(obj, replacer)).toEqual({ foo: 'Foo', deep: ['Bar'] })
  })

  it('patch all the props of an object with leaves that has nested array', () => {
    const obj = { foo: 'FOO', deep: [{ bar: 'BAR', deep: [{ foo: 'FOO' }] }] }
    expect(walker(obj, replacer)).toEqual({ foo: 'Foo', deep: [{ bar: 'Bar', deep: [{ foo: 'Foo' }] }] })
  })

  it('patch all the props of an object with leaves that has nested array', () => {
    const obj = { foo: 'FOO', deep: [{ bar: 'BAR', deep: [{ foo: 'FOO', bar: undefined }] }] }
    expect(walker(obj, replacer)).toEqual({
      foo: 'Foo',
      deep: [{ bar: 'Bar', deep: [{ foo: 'Foo', bar: undefined }] }]
    })
  })

  it('patch all the props of an object with leaves that has null values', () => {
    const obj = { foo: 'FOO', deep: [{ bar: 'BAR', deep: { foo: null, bar: null } }] }
    expect(walker(obj, replacer)).toEqual({
      foo: 'Foo',
      deep: [{ bar: 'Bar', deep: { foo: null, bar: null } }]
    })
  })
})
