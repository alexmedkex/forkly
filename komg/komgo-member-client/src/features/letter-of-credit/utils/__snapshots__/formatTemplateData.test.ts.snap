// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatTemplateData matches snapshot 1`] = `
Object {
  "amount": "100.00",
  "applicant": Object {
    "_id": "5d257a68e6280a7f7e2bfb68",
    "createdAt": "2019-07-09T00:00:00.000Z",
    "isFinancialInstitution": false,
    "isMember": true,
    "label": "0x4b8031c0bf1350534d733f16ff67080b5f208aa22e89b291e396b8176fb03436",
    "node": "0xa12b7e68c7bbc4e1f6080c96a7461d449d27e98c31020e4229d8cd8b4942009a",
    "owner": "0x734617Fc250f867C7E3EBa9E736fB8C9eD680cbf",
    "parentNode": "0x9c19ca3566862e19c7779f1af09f5b2883d8880917cedcbf4ae505a775fa518a",
    "staticId": "cf63c1f8-1165-4c94-a8f8-9252eb4f0016",
    "updatedAt": "2019-07-09T00:00:00.000Z",
    "x500Name": Object {
      "C": "country",
      "CN": "Applicant Name",
      "L": "city",
      "O": "Applicant Name ltd",
      "PC": "postal code",
      "STREET": "street",
    },
  },
  "beneficiary": Object {
    "_id": "5d257a68e6280a7f7e2bfb68",
    "createdAt": "2019-07-09T00:00:00.000Z",
    "isFinancialInstitution": false,
    "isMember": true,
    "label": "0x4b8031c0bf1350534d733f16ff67080b5f208aa22e89b291e396b8176fb03436",
    "node": "0xa12b7e68c7bbc4e1f6080c96a7461d449d27e98c31020e4229d8cd8b4942009a",
    "owner": "0x734617Fc250f867C7E3EBa9E736fB8C9eD680cbf",
    "parentNode": "0x9c19ca3566862e19c7779f1af09f5b2883d8880917cedcbf4ae505a775fa518a",
    "staticId": "ecc3b179-00bc-499c-a2f9-f8d1cc58e9db",
    "updatedAt": "2019-07-09T00:00:00.000Z",
    "x500Name": Object {
      "C": "country",
      "CN": "Beneficiary Name",
      "L": "city",
      "O": "Beneficiary Name ltd",
      "PC": "postal code",
      "STREET": "street",
    },
  },
  "beneficiaryBank": Object {
    "_id": "5d257a68e6280a7f7e2bfb68",
    "createdAt": "2019-07-09T00:00:00.000Z",
    "isFinancialInstitution": true,
    "isMember": true,
    "label": "0x4b8031c0bf1350534d733f16ff67080b5f208aa22e89b291e396b8176fb03436",
    "node": "0xa12b7e68c7bbc4e1f6080c96a7461d449d27e98c31020e4229d8cd8b4942009a",
    "owner": "0x734617Fc250f867C7E3EBa9E736fB8C9eD680cbf",
    "parentNode": "0x9c19ca3566862e19c7779f1af09f5b2883d8880917cedcbf4ae505a775fa518a",
    "staticId": "08e9f8e3-94e5-459e-8458-ab512bee6e2c",
    "updatedAt": "2019-07-09T00:00:00.000Z",
    "x500Name": Object {
      "C": "country",
      "CN": "Beneficiary Name",
      "L": "city",
      "O": "Beneficiary Name ltd",
      "PC": "postal code",
      "STREET": "street",
    },
  },
  "cargo": Object {
    "_id": "123",
    "cargoId": "F0401",
    "createdAt": "2019-01-01T00:00:00.000Z",
    "grade": "BRENT",
    "originOfGoods": "China",
    "parcels": Array [
      Object {
        "_id": undefined,
        "deemedBLDate": undefined,
        "destinationPlace": "Weymouth",
        "id": "123",
        "laycanPeriod": Object {
          "endDate": "2017-12-31",
          "startDate": "2017-12-31",
        },
        "loadingPlace": "Porto",
        "modeOfTransport": "VESSEL",
        "quantity": 600000,
        "version": 2,
      },
    ],
    "quality": "top",
    "source": "KOMGO",
    "sourceId": "E2389423",
    "status": "TO_BE_FINANCED",
    "updatedAt": "2019-01-01T00:00:00.000Z",
    "version": 2,
  },
  "currency": "USD",
  "expiryDate": "2019-08-21",
  "issuingBank": Object {
    "_id": "5d257a68e6280a7f7e2bfb68",
    "createdAt": "2019-07-09T00:00:00.000Z",
    "isFinancialInstitution": true,
    "isMember": true,
    "label": "0x4b8031c0bf1350534d733f16ff67080b5f208aa22e89b291e396b8176fb03436",
    "node": "0xa12b7e68c7bbc4e1f6080c96a7461d449d27e98c31020e4229d8cd8b4942009a",
    "owner": "0x734617Fc250f867C7E3EBa9E736fB8C9eD680cbf",
    "parentNode": "0x9c19ca3566862e19c7779f1af09f5b2883d8880917cedcbf4ae505a775fa518a",
    "staticId": "08e9f8e3-94e5-459e-8458-ab512bee6e2c",
    "updatedAt": "2019-07-09T00:00:00.000Z",
    "x500Name": Object {
      "C": "country",
      "CN": "Issuing Name",
      "L": "city",
      "O": "Issuing Name ltd",
      "PC": "postal code",
      "STREET": "street",
    },
  },
  "issuingBankReference": "REF-0001",
  "trade": Object {
    "_id": "123",
    "buyer": "node.123",
    "buyerEtrmId": "buyerEtrmId",
    "commodity": "Crude oil",
    "contractDate": "2017-12-01",
    "createdAt": "2019-01-01T00:00:00.000Z",
    "creditRequirement": "Documentary letter of credit",
    "currency": "EUR",
    "dealDate": "2017-12-31",
    "deliveryLocation": "Weymouth",
    "deliveryPeriod": Object {
      "endDate": "2019-04-10",
      "startDate": "2019-04-03",
    },
    "deliveryTerms": "FOB",
    "invoiceQuantity": "Load",
    "law": "New York law",
    "maxTolerance": 1.25,
    "minTolerance": 1.25,
    "paymentTerms": Object {
      "dayType": "Calendar",
      "eventBase": "BL",
      "time": 30,
      "timeUnit": "Days",
      "when": "After",
    },
    "paymentTermsOption": "Deferred",
    "price": "70.02",
    "priceUnit": "BBL",
    "quantity": 600000,
    "seller": "node.321",
    "sellerEtrmId": undefined,
    "source": "komgo",
    "sourceId": "E2389423",
    "status": "To be financed",
    "updatedAt": "2019-01-01T00:00:00.000Z",
    "version": 2,
  },
  "version": 1,
}
`;
